torchao/quantization/quantize_/workflows/int4/int4_tensor_core_tile_packed_tensor.py

```diff
@@ -0,0 +1,331 @@
+            (0, in_features - orig_in_features, 0, out_features - orig_out_features),
+        )
+
+        # Quantize
+        target_dtype = torch.int32
+        quant_min = 0
+        quant_max = 15
+
+        from torchao.quantization.quant_primitives import (
+            MappingType,
+            _choose_qparams_affine_tinygemm,
+            _quantize_affine_tinygemm,
+        )
+
+        # Calculate scale and zero_point for tinygemm
+        scale, zero_point = _choose_qparams_affine_tinygemm(
+            hp_tensor_padded,
+            mapping_type=MappingType.ASYMMETRIC,
+            block_size=tuple(block_size),
+            target_dtype=target_dtype,
+            quant_min=quant_min,
+            quant_max=quant_max,
+            scale_dtype=hp_tensor.dtype,
+            zero_point_dtype=hp_tensor.dtype,
+        )
+
+        # Quantize for tinygemm
+        int_data = _quantize_affine_tinygemm(
+            hp_tensor_padded,
+            block_size,
+            scale,
+            zero_point,
+            target_dtype,
+            quant_min=quant_min,
+            quant_max=quant_max,
+        )
+
+        # Convert to packed format
+        def quant_2d(int_data_2d):
+            int_data_2d = (int_data_2d[::, ::2] << 4 | int_data_2d[::, 1::2]).to(
+                torch.uint8
+            )
+            return torch.ops.aten._convert_weight_to_int4pack(
+                int_data_2d.contiguous(), inner_k_tiles
+            )
+
+        if int_data.dim() == 3:  # for moe quant
+            num_experts = int_data.shape[0]
+            packed_weight_list = []
+            for expert in range(num_experts):
+                packed_weight_list.append(quant_2d(int_data[expert]).unsqueeze(0))
+            packed_weight = torch.cat(packed_weight_list, dim=0)
+            scale = scale.reshape(int_data.shape[0], int_data.shape[-2], -1)
+            zero_point = (
+                zero_point.reshape(int_data.shape[0], int_data.shape[-2], -1)
+                if zero_point is not None
+                else None
+            )
+        else:
+            assert int_data.dim() == 2
+            packed_weight = quant_2d(int_data)
+            scale = scale.reshape(int_data.shape[0], -1)
+            zero_point = (
+                zero_point.reshape(int_data.shape[0], -1)
+                if zero_point is not None
+                else None
+            )
+
+        from torchao.quantization.utils import pack_tinygemm_scales_and_zeros
+
+        scale_and_zero = pack_tinygemm_scales_and_zeros(scale, zero_point, scale.dtype)
+
+        return cls(
+            qdata=packed_weight,
+            scale_and_zero=scale_and_zero,
+            block_size=block_size,
+            shape=original_shape,
+        )
+
+    def to(self, *args, **kwargs):
+        kwargs = self._get_to_kwargs(*args, **kwargs)
+        device = kwargs["device"]
+
+        # tensor core tiled layout supports both cpu and cuda but does not support the conversion
```

Comment:
sounds like it should be two separate layouts then, IMO "layout" means "how bytes are laid out in memory/on disk", and it 100% should be possible to copy arbitrary bytes between devices.  I think a different name is needed for what this PR is currently doing.

Reply:
yeah they do use different layouts (Int4CPULayout or something like that), this was trying to guard the device changes, since people can do:

in cpu: packed_weight = int4mm_pack(weight)
then packed_weight_cuda = packed_weight.to("cuda"), which would mess up the data entirely since `packed_weight_cuda` can not be generated this way.

> I think a different name is needed for what this PR is currently doing.

you mean just this comment or the Int4TensorCoreTilePackedTensor name should change

Reply:
I think can remove this, I think the issue is probably fixed now since we don't have CPU dispatch support for the weight packing op anymore: https://github.com/pytorch/pytorch/blob/5afa4187dfe1e99278f8e372ec09102d5b937572/aten/src/ATen/native/native_functions.yaml#L4203